[Chua Wei Jie Timothy] iP

ooawagaeri opened the pull request:
# CS2103/T Duke Project  

<img height="100" src="https://upload.wikimedia.org/wikipedia/commons/thumb/5/5d/Duke_%28Java_mascot%29_waving.svg/226px-Duke_%28Java_mascot%29_waving.svg.png"/>

## Duke Template

Based on a project template for a greenfield Java project. It's named after the [Java mascot _Duke_](https://www.oracle.com/java/duke.html).

- Project Duke is an educational software project designed to take you through the steps of building a small software incrementally while applying as many Java and SE techniques as possible along the way.
- The project aims to build a product named Duke, a Personal Assistant Chatbot that helps a person to keep track of various things.

Given below are instructions on how to use it:

## Setting up in Intellij :zap:

Prerequisites: JDK 11, update Intellij to the most recent version.

1. Open Intellij (if you are not in the welcome screen, click `File` > `Close Project` to close the existing project first)
2. Open the project into Intellij as follows:
   1. Click `Open`.
   2. Select the project directory, and click `OK`.
   3. If there are any further prompts, accept the defaults.
3. Configure the project to use **JDK 11** (not other versions) as explained in [here](https://www.jetbrains.com/help/idea/sdk.html#set-up-jdk)<br>
   In the same dialog, set the **Project language level** field to the `SDK default` option.
4. After that, locate the `src/main/java/Duke.java` file, right-click it, and choose `Run Duke.main()` (if the code editor is showing compile errors, try restarting the IDE). If the setup is correct, you should see something like the below as the output:
   ```
   Hello from
    ____        _        
   |  _ \ _   _| | _____ 
   | | | | | | | |/ / _ \
   | |_| | |_| |   <  __/
   |____/ \__,_|_|\_\___|
   ____________________________________________________________
   Hello! I'm Duke
   What can I do for you?
   ____________________________________________________________
   ```

## List of Features :bread:
- [X] Add Command
- [X] Delete Command
- [X] Done Command
- [X] List Command
- [X] Find Command
- [X] Done Command
- [ ] ~~Useless Command~~
- [ ] _More features to come!_

## How To Use :bulb:

> Prepare a `tasks.txt` containing a list of commands & task to be **initialized** <br>
> This file is _optional_ but is recommended to test the program's features

For example:
```text
add todo read book
add deadline return book /by 6/6/2021 1600
```

## Some Code :package:

```java
public static void main(String[] args) {
     new Duke("data/tasks.txt").run();
 }
 ```

Commit:
Add Gradle support

ooawagaeri committed:
Greet, Echo, Exit

ooawagaeri committed:
Add, List

ooawagaeri committed:
Mark as Done

ooawagaeri committed:
ToDos, Events, Deadlines

ooawagaeri committed:
Automated Text UI Testing

ooawagaeri committed:
Handle Errors

ooawagaeri committed:
Use Enums

ooawagaeri committed:
Save

ooawagaeri committed:
Dates and Times

ooawagaeri committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

ooawagaeri committed:
Save, Dates and Times Updates

ooawagaeri committed:
Package and More OOP: Command, TaskList, UI, Parser, Storage, Exception

ooawagaeri committed:
Amended Prints Of Commands and New JUnit

ooawagaeri committed:
Generated Artifact File for JAR

ooawagaeri committed:
Add JavaDoc To Duke Packages

ooawagaeri committed:
Rename Test Case Methods

ooawagaeri committed:
Add Find And Further OOP Commands From TaskList

ooawagaeri committed:
Merge branch 'branch-A-CodingStandard'

ooawagaeri committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	data/duke.txt
#	src/main/java/duke/storage/Storage.java
#	src/main/java/duke/tasklist/Task.java
#	src/main/java/duke/tasklist/TaskList.java
#	src/main/java/duke/ui/Ui.java

ooawagaeri committed:
Clean Up JavaDoc With New OOP Command Structure

ooawagaeri committed:
Update README.md For Markdown Requirement

ooawagaeri committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

marcuschj commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,46 @@
+ * Represents an addition user command.
+ * Inherits from command class.
+ */
+public class AddCommand extends Command {
```
Typically Class Names should be nouns, maybe change `AddCommand` to be `CommandAdder` instead? :)

ooawagaeri replied:
Noted! Will consider.

kjpvi replied:
same as Marcus's suggestion, nothing to add JYs!

marcuschj commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,46 @@
+    /**
+     * Returns if command exits program.
+     *
+     * @return if command exists program.
```
`@return false` would be more apt and indicative of the method

ooawagaeri replied:
Agreed, it would be clearer for readers.

marcuschj commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,52 @@
+ * Represents a deletion user command.
+ * Inherits from command class.
+ */
+public class DeleteCommand extends Command {
```
Same comment as before, maybe `CommandRemover` ?

marcuschj commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,52 @@
+    /**
+     * Returns if command exits program.
+     *
+     * @return if command exists program.
```
same as before, `@return false` would be better

marcuschj commented on src/main/java/duke/command/DoneCommand.java:
```diff
@@ -0,0 +1,52 @@
+    /**
+     * Returns if command exits program.
+     *
+     * @return if command exists program.
```
`@return false`

marcuschj commented on src/main/java/duke/command/ExitCommand.java:
```diff
@@ -0,0 +1,43 @@
+    /**
+     * Returns if command exits program.
+     *
+     * @return if command exists program.
```
`@return true`

marcuschj commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,43 @@
+ * Represents a find / query user command.
+ * Inherits from command class.
+ */
+public class FindCommand extends Command {
```
`CommandFinder`?

_Actually ... the class name is vague as it sounds like a verb (even though I know it was meant as a noun)_

ooawagaeri replied:
Yea honestly agree that its vague as "command" is a noun as well

marcuschj commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,43 @@
+    /**
+     * Returns if command Finds program.
+     *
+     * @return if command exists program.
```
`@return false`

marcuschj commented on src/main/java/duke/command/ListCommand.java:
```diff
@@ -0,0 +1,44 @@
+ * Represents a list user command.
+ * Inherits from command class.
+ */
+public class ListCommand extends Command {
```
`CommandLister`

marcuschj commented on src/main/java/duke/command/ListCommand.java:
```diff
@@ -0,0 +1,44 @@
+    /**
+     * Returns if command exits program.
+     *
+     * @return if command exists program.
```
`@return false`

marcuschj commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,56 @@
+            return new ExitCommand("");
+        } else if (cmd.contains("find")) {
+            return new FindCommand(cut(cmd, "find"));
+        }  else {
```
There is an extra space before `else`

ooawagaeri replied:
I see. Thanks for spotting the small detail.

marcuschj commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,71 @@
+        String location = "data/duke.txt";
+
+        try {
+
```
extra line

ooawagaeri replied:
Will be trimming.

marcuschj commented on src/main/java/duke/tasklist/Task.java:
```diff
@@ -0,0 +1,65 @@
+
+    /**
+     * Returns name of task.
+     */
```
missing `@return` javadoc

ooawagaeri replied:
Thanks for point it out, will amend.

marcuschj commented on src/main/java/duke/tasklist/TaskList.java:
```diff
@@ -0,0 +1,204 @@
+     */
+    private LocalDateTime dateTime(String time) throws DukeException {
+        DateTimeFormatter fmt = new DateTimeFormatterBuilder()
+                .appendPattern("d/M/yyyy")
+                .optionalStart()
+                .appendPattern(" HHmm")
+                .optionalEnd()
+                .parseDefaulting(ChronoField.HOUR_OF_DAY, 23)
+                .parseDefaulting(ChronoField.MINUTE_OF_HOUR, 59)
+                .toFormatter();
```
indentation should be 8, seems like it's more than that

ooawagaeri replied:
It is 8 from the previous line.

marcuschj commented on src/test/java/DukeTest.java:
```diff
@@ -0,0 +1,41 @@
+
```
Extra line

ooawagaeri replied:
Corrected

marcuschj commented on src/test/java/DukeTest.java:
```diff
@@ -0,0 +1,41 @@
+import duke.tasklist.Event;
+import duke.tasklist.Todo;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Static import should come first?

ooawagaeri replied:
Yup, I ran checkstyle to fix it.

marcuschj commented on src/test/java/DukeTest.java:
```diff
@@ -0,0 +1,41 @@
+
+import duke.tasklist.Deadline;
```
Package imports should come after

marcuschj commented on src/test/java/DukeTest.java:
```diff
@@ -0,0 +1,41 @@
+    @Test
+    public void test_create_deadline() {
+        Deadline dl = new Deadline("return book ", LocalDateTime.of(2021,
+                                        Month.JULY, 29, 12, 0, 0));
```
Wrong indentation

ooawagaeri replied:
Noted, fixed.

marcuschj commented on src/test/java/DukeTest.java:
```diff
@@ -0,0 +1,41 @@
+        assertEquals(dl.toString(), "[D][X] return book (by: 2021 Jul 29 12.00 PM)");
+        assertEquals(dl.save(), "D | 1 | return book | 2021 Jul 29 12.00 PM");
+    }
+    @Test
```
Should have line between methods

ooawagaeri replied:
Yup, corrected.

marcuschj commented on src/test/java/DukeTest.java:
```diff
@@ -0,0 +1,41 @@
+        assertEquals(ev.toString(), "[E][X] project meeting (at: 2021 Aug 2 2.00 PM)");
+        assertEquals(ev.save(), "E | 1 | project meeting | 2021 Aug 2 2.00 PM");
+    }
+    @Test
```
Should have line between methods

ooawagaeri replied:
Yup, corrected.

marcuschj reviewed (commented):
Hi Timothy! I'm Marcus. Did a review for standard coding practices. Apologies if I had to nitpick because the code was very well formatted so there wasn't much that was substantial which I could say. Well done! One last thing in general do keep an extra line after your last line of code in each file, I believe it will fail checkStyle without that.

All the best!

ooawagaeri committed:
Add Gradle XML Files, Fix Checkstyle

ooawagaeri committed:
Add Gradle XML Files, Fix Checkstyle

ooawagaeri committed:
Merge remote-tracking branch 'origin/branch-A-Gradle' into branch-A-Gradle

# Conflicts:
#	.gitignore

ooawagaeri committed:
Add Gradle XML Files, Fix Checkstyle

ooawagaeri committed:
Merge branch 'branch-A-Gradle' into branch-A-CheckStyle

ooawagaeri committed:
Migrate Duke to JavaFX String-Based

ooawagaeri committed:
Add Ui Documentation, Varargs To SaveToFile And Fix Dynamic Height Dialogs

ooawagaeri committed:
Update Code Block Indentation

ooawagaeri committed:
Merge branch 'Use_GFMD'

ooawagaeri committed:
Merge branch 'branch-Level-10'

kjpvi commented on src/main/java/duke/command/DoneCommand.java:
```diff
@@ -0,0 +1,55 @@
+ * Represents a done user command.
+ * Inherits from command class.
+ */
+public class DoneCommand extends Command {
```
Just to add from Marcus's Suggestion maybe consider a class CommandManager to manage all the commands and have an enum of different types of command instead?

ooawagaeri committed:
Update README.md Header

ooawagaeri committed:
Merge branch 'Use-GFMD'

timiditi commented on src/main/java/DialogBox.java:
```diff
@@ -0,0 +1,85 @@
+     * @return new dialog object.
+     */
+    public static DialogBox getDukeDialog(String text, Image img) {
+        var db = new DialogBox(text, img);
```
Do you want to specify the datatype of db in its declaration instead of using `var`?

timiditi commented on src/main/java/duke/tasklist/TaskList.java:
```diff
@@ -0,0 +1,205 @@
+     * Loads task request from text file into list.
+     *
+     * @param taskFile scanner object of text file.
+     */
```
Might want to specify when an exception in thrown in the javadocs

timiditi reviewed (commented):
LGTM! nice code, just a couple of tiny nitpicks with code style.

ooawagaeri committed:
Build Gradle Config: Change application name

Incorrect mainClassName for build.gradle

Previous mainClassName did not correctly navigate to project's main. JAR could not execute due to wrong mainClassName.

Do ensure all paths and packages are updated from the template.

ooawagaeri committed:
Task List Documentation: Add exception documentation

Missing @throws documentation.

Addition of exception documentation is more friendly for other programmers reviewing the code.

Ensuring standardized and complete documentation would aid in readability of program and better explain reasoning behind errors thrown for other programmers.

ooawagaeri committed:
Dialog Box Method Variable: Removal var data type

DialogBox class method getDukeDialog uses var datatype instead of the actual variable type.

Change var to explicit type. This allows program to anticipate actual DialogBox type instead of anything.

Let's standardize and remove the use of var because var obscures the actual type of the object. If initialized object does not return the desired object type then program may fail.

Refer to ths RedHat article when dealing with vars:
https://developers.redhat.com/blog/2018/05/25/simplify-local-variable-type-definition-using-the-java-10-var-keyword

ooawagaeri committed:
Add Assertions on Duke, Parser, TaskList

ooawagaeri committed:
Fix Code Quality: Command, Cutter, Duke, Ui, Parser, TaskList

Minor coding quality issues such as code duplication, dead code and SLAP were corrected.

ooawagaeri committed:
Add Stream to TaskList Find

ooawagaeri committed:
Merge pull request #1 from ooawagaeri/branch-A-Assertions

Add Assertions on Duke, Parser, TaskList

ooawagaeri committed:
Merge branch 'master' into branch-A-CodeQuality

ooawagaeri committed:
Merge pull request #2 from ooawagaeri/branch-A-CodeQuality

Fix Code Quality: Command, Cutter, Duke, Ui, Parser, TaskList

ooawagaeri committed:
Fix Minor Conflict with getResponse()

ooawagaeri committed:
Add GitHub Actions CI

ooawagaeri committed:
Fix Minor Conflict with getResponse()

ooawagaeri committed:
Update gradle.yml

ooawagaeri committed:
Remove IO Test gradle.yml

ooawagaeri committed:
Add Sort Command

ooawagaeri committed:
Fix CheckStyle: Parser, SortCommand

ooawagaeri committed:
Merge remote-tracking branch 'origin/branch-A-CI'

ooawagaeri committed:
Add Background and Dialog Pointer

ooawagaeri committed:
Fix DialogBox size and Add User Guide

ooawagaeri committed:
Set theme jekyll-theme-slate

ooawagaeri committed:
Update ReadMe

ooawagaeri committed:
Set theme jekyll-theme-cayman

ooawagaeri committed:
Update ReadMe

ooawagaeri committed:
Fix Size Increment / Decrement
